File: solutions/go/strain/1/strain.go
package strain

type Ints []int
type Lists [][]int
type Strings []string

func (i Ints) Keep(filter func(int) bool) (output Ints) {
	for _, number := range i {
		if filter(number) {
			output = append(output, number)
		}
	}
	return
}

func (i Ints) Discard(filter func(int) bool) Ints {
	return i.Keep(func(number int) bool { return !filter(number) })
}

func (l Lists) Keep(filter func([]int) bool) (output Lists) {
	for _, number := range l {
		if filter(number) {
			output = append(output, number)
		}
	}
	return
}

func (s Strings) Keep(filter func(string) bool) (output Strings) {
	for _, word := range s {
		if filter(word) {
			output = append(output, word)
		}
	}
	return
}

File: solutions/javascript/strain/1/strain.js
/**
 * Filters an array based on a provided filter function and a boolean flag.
 *
 * @param {Array} array - The array to filter.
 * @param {Function} filter - The filter function to apply to each element.
 * @param {boolean} keepMatches - If true, keeps elements that match the filter; otherwise, discards them.
 * @returns {Array} The filtered array.
 */
const strain = (array, filter, keepMatches) => {
  const results = [];
  array.forEach(item => {
    if (filter(item) === keepMatches) {
      results.push(item);
    }
  });
  return results;
};

/**
 * Keeps elements in an array that match the provided filter function.
 *
 * @param {Array} array - The array to filter.
 * @param {Function} filter - The filter function to apply to each element.
 * @returns {Array} The array with matching elements kept.
 */
export const keep = (array, filter) => strain(array, filter, true);

/**
 * Discards elements in an array that match the provided filter function.
 *
 * @param {Array} array - The array to filter.
 * @param {Function} filter - The filter function to apply to each element.
 * @returns {Array} The array with matching elements discarded.
 */
export const discard = (array, filter) => strain(array, filter, false);
